lib/navigation/preprocess_encoders.hpp

```diff
@@ -1,27 +1,96 @@
+
 class EncodersPreprocessor {
  public:
-  EncodersPreprocessor();
+  EncodersPreprocessor(core::ILogger &logger);
 
-  core::EncoderData processData(const core::EncoderData encoder_data);
+  /**
+   * @brief process encoder data to deal with outliers and unreliable sensors
+   *
+   * @param encoder_data
+   * @return cleaned and reliable encoder data
+   */
+  std::optional<core::EncoderData> processData(const core::EncoderData &encoder_data);
 
  private:
-  core::EncoderData detectOutliers(const core::EncoderData encoder_data);
+  struct Statistics {
+    core::Float median;
+    core::Float upper_bound;
+    core::Float lower_bound;
+  };
+
+  std::optional<Statistics> getStatistics(const core::EncoderData &encoder_data) const;
+
+  /**
+   * @brief detects new outliers and replaces values from faulty sensors with
+   * the median; fails if not enough reliable data is available
+   *
+   * @param encoder_data to be sanitised
+   * @return consistent data with no outliers, if possible
+   */
+  std::optional<core::EncoderData> sanitise(const core::EncoderData &encoder_data);
+
+  /**
+   * @brief changes the value corresponding to the encoder sensor in the encoders_reliable_ array to
+   * false if the number of consecutive outliers exceeds the threshold value
+   *
+   * @return SensorChecks::kUnacceptable if more than 1 sensors have become unreliable and
+   * SensorChecks::kAcceptable otherwise
+   */
+  SensorChecks checkReliable();
 
-  void checkReliable(const core::EncoderData &encoder_data);
+  /**
+   * @brief calculates a specific quantile value from sorted input array
+   *
+   * @param reliable_data sorted array of values
+   * @param fraction in [0, 1] corresponding to the quantile (e.g. 0.25 for the 25th percentile)
+   * @return the quantile that we require
+   */
+  template<std::size_t N>
+  core::Float getSpecificQuantile(const std::array<std::uint32_t, N> &reliable_data,
+                                  const core::Float fraction) const
+  {
+    const core::Float theoretical_index = (num_reliable_encoders_ - 1) * fraction;
+    const std::size_t low_index         = static_cast<std::size_t>(std::floor(theoretical_index));
+    const std::size_t high_index        = static_cast<std::size_t>(std::ceil(theoretical_index));
+    const core::Float low               = static_cast<core::Float>(reliable_data.at(low_index));
+    const core::Float high              = static_cast<core::Float>(reliable_data.at(high_index));
+    return (low + high) / 2.0;
+  }
 
-  // initialised as {0, 0, 0, 0}, count of consecutive outliers
-  std::array<std::uint16_t, core::kNumEncoders> num_outliers_per_encoder_;
+  /**
+   * @brief cacluates the quartiles (q1,median,q3)
+   *
+   * @param array containing filtered values(i.e with no values from faulty sensors)
+   * @return a quartile object which contains q1, q3 and the median value
+   */
+  template<std::size_t N>
+  Quartile getQuartiles(std::array<std::uint32_t, N> &reliable_data) const
```

Comment:
since we know this will only run when N=4 or when N=3, why not explicitly instantiate this in the .hpp for quicker compilation. (look [here](https://stackoverflow.com/questions/13068993/when-would-you-use-template-explicit-instantiation#:~:text=func%3Cdouble%3E()%3B%20//%20OK%0A%20%20%20%20func%3Cint%3E()%3B%20//%20OK) - don't follow the .cpp definition part here but the explicit instantiation is pretty useful)

Reply:
We were looking to make this as general as possible for future teams, so this function could run with any number of sensors